klippy/extras/display/menu.py

```diff
@@ -0,0 +1,1137 @@
+
+    def populate_items(self):
+        super(MenuVSDCard, self).populate_items()
+        self._populate_files()
+
+class MenuCard(MenuGroup):
+    def __init__(self, menu, config):
+        super(MenuCard, self).__init__(menu, config)
+        self.content = config.get('content')
+
+    def _names_aslist(self):
+        return self._aslist_splitlines(self.items)
+
+    def _content_aslist(self):
+        return filter(None, [self._strip_quotes(s) for s in self._aslist_splitlines(self.content)])
+
+    def update_items(self):
+        self._items = self._allitems[:]
+        for item in self._items:
+            if isinstance(item, MenuGroup) and not item.is_editing():
+                item.update_items()
+
+    def _lookup_item(self, item):
+        if isinstance(item, str) and ',' in item:
+            item = MenuCycler(self._menu, {'name':'Cycler', 'items': item}, ',')
+        return super(MenuCard, self)._lookup_item(item)
+
+    def render_content(self, eventtime):
+        if self.selected is not None:
+            self.selected = (self.selected % len(self)) if len(self) > 0 else None
+
+        items = []
+        for i, item in enumerate(self):
+            name = ''
+            if item.is_enabled():
+                item.heartbeat(eventtime)
+                name = self._render_item(item, (i == self.selected), True)
+            items.append(name)
+        lines = []
+        for line in self._content_aslist():
+            try:
+                lines.append(str(line).format(*items))
+            except Exception as e:
+                logging.error('Card rendering exception: '+ str(e))
+        return lines
+
+    def _render(self):
+        return self._name
+
+class MenuDeck(MenuList):
+    def __init__(self, menu, config):
+        super(MenuDeck, self).__init__(menu, config)
+        self._show_back = False
+        self._show_title = False
+
+    def _names_aslist(self):
+        return self._aslist(self.items)
+
+    def is_accepted(self, item):
+        return type(item) is MenuCard
+
+    def _render(self):
+        return self._name
+
+menu_items = { 'item': MenuItem, 'command': MenuCommand, 'input': MenuInput, 'list': MenuList, 'vsdcard':MenuVSDCard, 'deck':MenuDeck, 'card': MenuCard }
+# Default dimensions for lcds (rows, cols)
+LCD_dims = { 'st7920': (4,16), 'hd44780': (4,20), 'uc1701' : (4,16) }
+
+TIMER_DELAY   = 0.200
+BLINK_FAST_SEQUENCE = (True, True, False, False)
+BLINK_SLOW_SEQUENCE = (True, True, True, True, False, False, False)
+
+class MenuManager:
+    def __init__(self, config):
+        self.running = False
+        self.menuitems = {}
+        self.menustack = []
+        self._recursive_guard = []
+        self.top_row = 0
+        self.selected = 0
+        self.blink_fast_state = True
+        self.blink_slow_state = True
+        self.blink_fast_idx = 0
+        self.blink_slow_idx = 0
+        self.timeout_idx = 0
+        self.config = config
```

Comment:
The code should not store a reference to the config object past the config file reading phase.  (Otherwise, it can lead to confusion with config error reporting.) The self.config.get_prefix_sections() usage later in the code can be replaced with self.printer.lookup_module_objects() (or a similar call).